File: sdk/extension/transport.go
package extension

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
)

const (
	initialBufferSize = 1024 * 1024
	// MaxMessageSize bounds one encoded JSON-RPC message.
	MaxMessageSize = 10 * 1024 * 1024
)

// Message is one line-delimited JSON-RPC envelope.
type Message struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *Error          `json:"error,omitempty"`
}

// Error is the public JSON-RPC error object returned by the extension or host.
type Error struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data,omitempty"`
}

// Error returns a stable human-readable message.
func (e *Error) Error() string {
	if e == nil {
		return "<nil>"
	}
	if len(bytes.TrimSpace(e.Data)) == 0 {
		return fmt.Sprintf("code %d: %s", e.Code, e.Message)
	}
	return fmt.Sprintf("code %d: %s (%s)", e.Code, e.Message, string(e.Data))
}

// DecodeData unmarshals the structured error data into target.
func (e *Error) DecodeData(target any) error {
	if e == nil {
		return fmt.Errorf("decode rpc error data: missing error")
	}
	if len(bytes.TrimSpace(e.Data)) == 0 {
		return io.EOF
	}
	return json.Unmarshal(e.Data, target)
}

// Transport exchanges line-delimited JSON-RPC messages.
type Transport interface {
	ReadMessage() (Message, error)
	WriteMessage(Message) error
	Close() error
}

// StdIOTransport implements the line-delimited subprocess transport used by
// executable extensions.
type StdIOTransport struct {
	scanner *bufio.Scanner
	reader  io.Closer
	writer  io.Closer
	out     io.Writer
	mu      sync.Mutex
}

// NewStdIOTransport constructs a line-delimited transport over reader/writer.
func NewStdIOTransport(reader io.Reader, writer io.Writer) *StdIOTransport {
	scanner := bufio.NewScanner(reader)
	buffer := make([]byte, 0, initialBufferSize)
	scanner.Buffer(buffer, MaxMessageSize)

	transport := &StdIOTransport{
		scanner: scanner,
		out:     writer,
	}
	if closer, ok := reader.(io.Closer); ok {
		transport.reader = closer
	}
	if closer, ok := writer.(io.Closer); ok {
		transport.writer = closer
	}
	return transport
}

// ReadMessage reads the next non-empty message from the transport.
func (t *StdIOTransport) ReadMessage() (Message, error) {
	if t == nil || t.scanner == nil {
		return Message{}, fmt.Errorf("read transport message: missing scanner")
	}

	for t.scanner.Scan() {
		line := bytes.TrimSpace(t.scanner.Bytes())
		if len(line) == 0 {
			continue
		}

		var message Message
		if err := json.Unmarshal(line, &message); err != nil {
			return Message{}, newParseError(map[string]any{"error": err.Error()})
		}
		return message, nil
	}

	if err := t.scanner.Err(); err != nil {
		if errors.Is(err, bufio.ErrTooLong) {
			return Message{}, newInternalError(map[string]any{"reason": "message_too_large"})
		}
		return Message{}, err
	}
	return Message{}, io.EOF
}

// WriteMessage writes one JSON-RPC message with a trailing newline.
func (t *StdIOTransport) WriteMessage(message Message) error {
	if t == nil || t.out == nil {
		return fmt.Errorf("write transport message: missing writer")
	}

	t.mu.Lock()
	defer t.mu.Unlock()

	message.JSONRPC = "2.0"
	encoded, err := json.Marshal(message)
	if err != nil {
		return err
	}
	encoded = append(encoded, '\n')
	_, err = t.out.Write(encoded)
	return err
}

// Close closes the underlying reader and writer when possible.
func (t *StdIOTransport) Close() error {
	if t == nil {
		return nil
	}

	var closeErr error
	if t.reader != nil {
		closeErr = errors.Join(closeErr, t.reader.Close())
	}
	if t.writer != nil && t.writer != t.reader {
		closeErr = errors.Join(closeErr, t.writer.Close())
	}
	return closeErr
}

func marshalJSON(value any) (json.RawMessage, error) {
	if value == nil {
		return json.RawMessage("{}"), nil
	}
	encoded, err := json.Marshal(value)
	if err != nil {
		return nil, err
	}
	return encoded, nil
}

func unmarshalJSON(raw json.RawMessage, target any) error {
	trimmed := bytes.TrimSpace(raw)
	if len(trimmed) == 0 || bytes.Equal(trimmed, []byte("null")) {
		return nil
	}
	return json.Unmarshal(trimmed, target)
}

func contextError(ctx context.Context) error {
	if ctx == nil {
		return nil
	}
	if cause := context.Cause(ctx); cause != nil {
		return cause
	}
	return ctx.Err()
}

func newParseError(data any) *Error {
	return newRPCError(-32700, "Parse error", data)
}

func newInvalidRequestError(data any) *Error {
	return newRPCError(-32600, "Invalid request", data)
}

func newMethodNotFoundError(method string) *Error {
	return newRPCError(-32601, "Method not found", map[string]any{"method": method})
}

func newInvalidParamsError(data any) *Error {
	return newRPCError(-32602, "Invalid params", data)
}

func newInternalError(data any) *Error {
	return newRPCError(-32603, "Internal error", data)
}

func newRPCError(code int, message string, data any) *Error {
	raw, err := marshalJSON(data)
	if err != nil {
		raw = nil
	}
	return &Error{Code: code, Message: message, Data: raw}
}

File: sdk/extension-sdk-ts/src/transport.ts
import { MAX_MESSAGE_SIZE } from "./types.js";

/** Unique identifier for one JSON-RPC message. */
export type MessageID = number | string;

/** Wire-format shape of a JSON-RPC error object. */
export interface RPCErrorShape {
  code: number;
  message: string;
  data?: unknown;
}

/** One JSON-RPC 2.0 message envelope. */
export interface Message {
  jsonrpc?: "2.0";
  id?: MessageID;
  method?: string;
  params?: unknown;
  result?: unknown;
  error?: RPCErrorShape;
}

/** Bidirectional message transport for the extension subprocess protocol. */
export interface Transport {
  readMessage(): Promise<Message>;
  writeMessage(message: Message): Promise<void>;
  close(): Promise<void>;
}

/** Signals that the transport stream has ended. */
export class EOFError extends Error {
  constructor(message = "end of stream") {
    super(message);
    this.name = "EOFError";
  }
}

/** Typed JSON-RPC error with code, message, and optional data. */
export class RPCError extends Error {
  readonly code: number;
  readonly data?: unknown;

  constructor(code: number, message: string, data?: unknown) {
    super(message);
    this.name = "RPCError";
    this.code = code;
    this.data = data;
  }

  /** Constructs an RPCError from a wire-format error shape. */
  static fromShape(shape: RPCErrorShape): RPCError {
    return new RPCError(shape.code, shape.message, shape.data);
  }

  /** Returns the wire-format representation of this error. */
  toShape(): RPCErrorShape {
    return { code: this.code, message: this.message, data: this.data };
  }

  /** Unmarshals the structured error data into the target type. */
  decodeData<T>(): T {
    return this.data as T;
  }
}

type ReadWaiter = {
  reject: (error: unknown) => void;
  resolve: (message: Message) => void;
};

/** Default transport reading line-delimited JSON from stdin and writing to stdout. */
export class StdIOTransport implements Transport {
  readonly input: NodeJS.ReadableStream;
  readonly output: NodeJS.WritableStream;

  private readonly queue: Message[] = [];
  private readonly waiters: ReadWaiter[] = [];
  private buffer = "";
  private closed = false;
  private closedError?: unknown;
  private closePromise?: Promise<void>;
  private readonly onData: (chunk: string | Buffer) => void;
  private readonly onEnd: () => void;
  private readonly onError: (error: Error) => void;

  /** Constructs a line-delimited transport over the given readable and writable streams. */
  constructor(
    input: NodeJS.ReadableStream = process.stdin,
    output: NodeJS.WritableStream = process.stdout
  ) {
    this.input = input;
    this.output = output;

    if ("setEncoding" in input && typeof input.setEncoding === "function") {
      input.setEncoding("utf8");
    }

    this.onData = chunk => {
      this.buffer += typeof chunk === "string" ? chunk : chunk.toString("utf8");
      this.drainBuffer();
    };
    this.onEnd = () => {
      this.closed = true;
      this.closedError = new EOFError();
      this.flushWaiters(this.closedError);
    };
    this.onError = error => {
      this.closed = true;
      this.closedError = error;
      this.flushWaiters(error);
    };

    input.on("data", this.onData);
    input.on("end", this.onEnd);
    input.on("error", this.onError);
  }

  /** Reads the next non-empty message from the transport. */
  async readMessage(): Promise<Message> {
    if (this.queue.length > 0) {
      const message = this.queue.shift();
      if (message === undefined) {
        throw new EOFError();
      }
      return message;
    }

    if (this.closed) {
      throw (this.closedError as Error | undefined) ?? new EOFError();
    }

    return new Promise<Message>((resolve, reject) => {
      this.waiters.push({ resolve, reject });
    });
  }

  /** Writes one JSON-RPC message with a trailing newline. */
  async writeMessage(message: Message): Promise<void> {
    if (this.closed) {
      throw new EOFError();
    }

    const envelope = { ...message, jsonrpc: "2.0" as const };
    const encoded = JSON.stringify(envelope);
    if (Buffer.byteLength(encoded, "utf8") > MAX_MESSAGE_SIZE) {
      throw newInternalError({ reason: "message_too_large" });
    }

    await new Promise<void>((resolve, reject) => {
      this.output.write(`${encoded}\n`, error => {
        if (error) {
          reject(error);
          return;
        }
        resolve();
      });
    });
  }

  /** Closes the underlying reader and writer and rejects any pending waiters. */
  async close(): Promise<void> {
    if (this.closePromise !== undefined) {
      return this.closePromise;
    }

    this.closed = true;
    this.closePromise = Promise.resolve().then(async () => {
      this.input.off("data", this.onData);
      this.input.off("end", this.onEnd);
      this.input.off("error", this.onError);
      this.closedError ??= new EOFError();
      this.flushWaiters(this.closedError);
    });
    return this.closePromise;
  }

  private drainBuffer(): void {
    while (true) {
      const newlineIndex = this.buffer.indexOf("\n");
      if (newlineIndex === -1) {
        if (Buffer.byteLength(this.buffer, "utf8") > MAX_MESSAGE_SIZE) {
          const error = newInternalError({ reason: "message_too_large" });
          this.closed = true;
          this.closedError = error;
          this.flushWaiters(error);
        }
        return;
      }

      const line = this.buffer.slice(0, newlineIndex);
      this.buffer = this.buffer.slice(newlineIndex + 1);
      if (Buffer.byteLength(line, "utf8") > MAX_MESSAGE_SIZE) {
        const error = newInternalError({ reason: "message_too_large" });
        this.closed = true;
        this.closedError = error;
        this.flushWaiters(error);
        return;
      }

      const trimmed = line.trim();
      if (trimmed.length === 0) {
        continue;
      }

      let message: Message;
      try {
        message = JSON.parse(trimmed) as Message;
      } catch (error) {
        this.closed = true;
        this.closedError = newParseError({ error: stringifyError(error) });
        this.flushWaiters(this.closedError);
        return;
      }

      const waiter = this.waiters.shift();
      if (waiter !== undefined) {
        waiter.resolve(message);
        continue;
      }

      this.queue.push(message);
    }
  }

  private flushWaiters(error: unknown): void {
    while (this.waiters.length > 0) {
      const waiter = this.waiters.shift();
      waiter?.reject(error);
    }
  }
}

/** Returns true if the given value is an RPCError instance. */
export function isRPCError(error: unknown): error is RPCError {
  return error instanceof RPCError;
}

/** Creates a JSON-RPC Parse error (-32700). */
export function newParseError(data?: unknown): RPCError {
  return new RPCError(-32700, "Parse error", data);
}

/** Creates a JSON-RPC Invalid Request error (-32600). */
export function newInvalidRequestError(data?: unknown): RPCError {
  return new RPCError(-32600, "Invalid request", data);
}

/** Creates a JSON-RPC Method Not Found error (-32601). */
export function newMethodNotFoundError(method: string): RPCError {
  return new RPCError(-32601, "Method not found", { method });
}

/** Creates a JSON-RPC Invalid Params error (-32602). */
export function newInvalidParamsError(data?: unknown): RPCError {
  return new RPCError(-32602, "Invalid params", data);
}

/** Creates a JSON-RPC Internal error (-32603). */
export function newInternalError(data?: unknown): RPCError {
  return new RPCError(-32603, "Internal error", data);
}

/** Converts a numeric or string message ID to its canonical string form. */
export function normalizeMessageID(id: MessageID): string {
  return String(id);
}

function stringifyError(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  return String(error);
}
